<a href="https://colab.research.google.com/github/Skapooky/AtividadePraticaRA2/blob/main/C%C3%B3pia_de_Atividade_Pr%C3%A1tica_Individual_RA2_2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Atenção: Regras para a solução da atividade.** 
1. copie este caderno para a sua conta, criando as células de texto e código que forem necessárias após o enunciado de cada questão;
2. não apague os enunciados; 
3. responda as questões na ordem em que elas se apresentam neste caderno;
4. pode, e deve, consultar todo e qualquer recurso disponível na internet. Mas, não pode consultar nenhum outro ser humano. 
5. Poste no AVA (Canvas) o link para a sua resposta. Pode ser um link para a solução no Google Colab, ou um link para o caderno com suas respostas no Github.

A violação de qualquer uma destas regras provoca a anulação da avaliação. 

Qualquer link que não possa ser executado a partir do link postado irá zerar a avaliação. 

Qualquer link que leve a uma pasta, arquivo comprimido, ou qualquer outro destino que não seja um caderno executável com as suas repostas provocará a anulação da questão. Teste usando outro navegador que não esteja logado nas suas contas. 

Qualquer link que não possa ser aberto, irá zerar a avaliação. 

Qualquer comunicação com outro ser humano irá zerar a avaliação.  

# Questão 1 (vale 6 pontos)
Considerando as funções listadas a seguir faça: **(a)** o gráfico das três funções; **(b)** o passo a passo algébrico, em latex, para encontrar o ângulo de inclinação da reta tangente no ponto $x=5$ para cada das funções:

1. $
f(x) = (x^2 - 4)(x^3 - 7x + 1)
$
2. $
f(x) = \sin(x^2)\cos(2x)
$
3.$
f(x) = x^2e^{-x}\ln(x)
$

Se você usar o sympy, terá que fazer com que ele imprima todos os passos necessários para resolver cada uma das operações, regras, ou tabelas que aplicar para resolver o problema.

Em todas as questões você pode usar apenas os métodos, regras, técnicas e algoritmos que utilizamos na nossa disciplina, em sala, ou no material de apoio.


Aluno: Henrique Anderle Schulz

In [53]:
!pip install plotly

import numpy as np
import sympy as sy
import plotly.graph_objects as go
import matplotlib.pyplot as plt
from mpmath import * 

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [65]:
#Definindo funçao 1
def f1(x):
  return ((x**2)-4)*(x**3 - 7*x + 1)

def f2(x):
  return np.sin(x**2)*np.cos(2*x)

def f3(x):
  return ((x**2)*np.e**-x*np.log(x))

xValues = np.arange(1, 100, 0.1)

fig1 = go.Figure()
fig1.add_trace(go.Scatter(x=xValues, y=f1(xValues), mode='lines'))
fig1.show()

fig2 = go.Figure()
fig2.add_trace(go.Scatter(x=xValues, y=f2(xValues), mode='lines'))
fig2.show()

fig3 = go.Figure()
fig3.add_trace(go.Scatter(x=xValues, y=f3(xValues), mode='lines'))
fig3.show()

print(f1(5))
print(f2(5))
print(f3(5))

print(f1(5.1))
print(f2(5.1))
print(f3(5.1))

1911
0.11105258533048293
0.27110768380749317
2155.9015099999997
-0.5492627211062305
0.2583590639566352


Função 1
$$ \lim\limits_{b \to a} = \frac{f(b) - f(a)}{b - a}$$

$$ \lim\limits_{b \to a} = \frac{2156 - 1911}{5.1 - 5}$$

$$ \lim\limits_{b \to a} = \frac{245}{0.1}$$

$$ \lim\limits_{b \to a} = 2415$$

$$ \lim\limits_{b \to a} = 255°$$

Função 2

$$ \lim\limits_{b \to a} = \frac{f(b) - f(a)}{b - a}$$

$$ \lim\limits_{b \to a} = \frac{-0.55 - 0.11.}{5.1 - 5}$$

$$ \lim\limits_{b \to a} = \frac{-0.44}{0.1}$$

$$ \lim\limits_{b \to a} = -4.4°$$

Função 3

$$ \lim\limits_{b \to a} = \frac{f(b) - f(a)}{b - a}$$

$$ \lim\limits_{b \to a} = \frac{0.25 - 0.27.}{5.1 - 5}$$

$$ \lim\limits_{b \to a} = \frac{-0.02}{0.1}$$

$$ \lim\limits_{b \to a} = -0.2°$$

# Questão 2 (vale 2 pontos)
Uma empresa espacial está tentando descobrir o ponto mínimo da órbita de um satélite cuja função orbital é dada por: 
$$f(x)=x^4−5x^3+10x^2−10x+5$$

Encontre o primeiro ponto de mínimo no intervalo $-10<=x<=10$.

Como sempre, você pode usar qualquer ferramenta disponível na internet para lhe auxiliar a resolver este problema. Lembre-se, contudo, que o sympy é uma boa alternativa.

Em todas as questões você pode usar apenas os métodos, regras, técnicas e algoritmos que utilizamos na nossa disciplina, em sala, ou no material de apoio. Lembrem-se estamos limitados as bibliotecas sympy, numpy e matplotlib.

Para resolver esta questão você precisará de, no minimo, 4 casas decimais de precisão depois da vírgula. 

In [41]:
#Definindo precisão
mp.prec = 100
mp.dps = 4

#Definido f(x)
def g(x):
  return (x**4) - (5*x**3) + (10*x**2) - (10*x +5)

#Definindo segunda derivada
def derivada2(x):
  return 8*x**2 - 30*x + 20

#Criando vetor de x
xValues = np.linspace(-11, 11, 1000)

#Plotando Gráfico
fig = go.Figure()
fig.add_trace(go.Scatter(x=xValues, y=g(xValues), mode='lines'))
fig.show()

#Achar pontos críticos algébricos
x = sy.Symbol('x')
f_x = (x**mpf(4)) - (mpf(5)*x**mpf(3)) + (mpf(10)*x**mpf(2)) - (mpf(10)*x +mpf(5))

f_prime_x = sy.diff(f_x, x)
pontos_criticos = sy.solve(f_prime_x, x)

#print(derivada2(pontos_criticos[0])) é menor que 0

print(f"Primeiro ponto mínimo: {pontos_criticos[0]}")


Primeiro ponto minimo: 1.60582958618827


# Questão 3 (vale 2 pontos)
Considerando a função $f(x,y) = \sin(xy) + \cos(x+y)$, no intervalo $0\leq x\leq 5 \space\space e \space\space 0\leq y\leq 5$ trace, usando o plotly o gráfico desta função e marque os pontos de máximo e mínimo neste intevalo.

Lembre-se o sympy é sempre uma opção e você pode usar qualquer técnica, algoritmo, regra, ou tabela, que tenha sido usada na matéria. E que precisa de apenas uma casa decimal de precisão. Uma dica importante para a solução deste problema é a função **np.unravel_index**


In [77]:
def f(x, y):
    return np.sin(x*y) + np.cos(x+y)

x = np.linspace(0, 5, 100)
y = np.linspace(0, 5, 100)

X, Y = np.meshgrid(x, y)
Z = f(X, Y)

fig1 = go.Figure(data=[go.Surface(x=X, y=Y, z=Z)])
fig1.update_layout(title="Gráfico de f(x,y) = sin(xy) + cos(x+y)", autosize=False, width=600, height=600)
fig1.show()